flink-core/src/main/java/org/apache/flink/core/fs/DuplicatingFileSystem.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.core.fs;
+
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * An extension interface for {@link FileSystem FileSystems} that can perform cheap DFS side
+ * duplicate operation. Such an operation can improve the time required for creating cheaply
+ * independent snapshots from incremental snapshots.
+ */
+public interface DuplicatingFileSystem {
```

Comment:
Right now, we get FileSystem from Path#getFileSystem. Do I understand correctly that we will need to rewrite this code as well?

Reply:
Not sure, if I understand your concern. Do you mind to elaborate?

Reply:
My question was how does this interface will be used. I thought that we are using FileSystem implementation from a third-party library but now I understand we have our own implementation as well so it should not be a problem to expand the such implementation with this interface.